File: Assets/Scripts/Match/Referee.cs
using System;
using System.Collections.Generic;

namespace FIF.Match
{
    /// <summary>How play is restarted after the ball leaves play or an infringement.</summary>
    public enum RestartType
    {
        KickOff,
        ThrowIn,
        GoalKick,
        Corner,
        FreeKick,
        Penalty
    }

    /// <summary>A pending restart: where the ball goes and who takes it.</summary>
    public readonly struct Restart
    {
        public readonly RestartType type;
        public readonly Vec2 ballPosition;
        public readonly TeamSide takingSide;

        public Restart(RestartType type, Vec2 ballPosition, TeamSide takingSide)
        {
            this.type = type;
            this.ballPosition = ballPosition;
            this.takingSide = takingSide;
        }
    }

    /// <summary>
    /// The laws layer of the headless match core (Part D): detects goals, decides
    /// how the ball comes back into play when it goes out, judges offside, and
    /// awards free kicks/penalties for fouls. Pure functions over the pitch and
    /// positions so the engine stays the single source of truth for state.
    /// </summary>
    public static class Referee
    {
        public static TeamSide Other(TeamSide side) =>
            side == TeamSide.Home ? TeamSide.Away : TeamSide.Home;

        /// <summary>
        /// Returns true if the ball has fully crossed a goal line between the
        /// posts. <paramref name="scoringSide"/> is the side that scored.
        /// </summary>
        public static bool IsGoal(Pitch pitch, Vec2 ball, out TeamSide scoringSide)
        {
            float halfGoal = pitch.GoalWidth * 0.5f;
            if (Math.Abs(ball.x) <= halfGoal)
            {
                if (ball.y >= pitch.HalfLength)
                {
                    // Crossed the +y line: Home attacks +y, so Home scored.
                    scoringSide = TeamSide.Home;
                    return true;
                }
                if (ball.y <= -pitch.HalfLength)
                {
                    scoringSide = TeamSide.Away;
                    return true;
                }
            }
            scoringSide = TeamSide.Home;
            return false;
        }

        /// <summary>Kick-off restart for the given side at the centre spot.</summary>
        public static Restart KickOff(Pitch pitch, TeamSide takingSide) =>
            new Restart(RestartType.KickOff, pitch.Centre, takingSide);

        /// <summary>
        /// Resolves a ball that is out of play (already known to be out and not a
        /// goal). <paramref name="lastTouch"/> is the side that last played it.
        /// </summary>
        public static Restart ResolveBoundary(Pitch pitch, Vec2 ball, TeamSide lastTouch)
        {
            TeamSide opp = Other(lastTouch);

            // Side line → throw-in to the team that did not touch it last.
            if (Math.Abs(ball.x) > pitch.HalfWidth)
            {
                float clampedY = Clamp(ball.y, -pitch.HalfLength, pitch.HalfLength);
                float x = ball.x > 0f ? pitch.HalfWidth : -pitch.HalfWidth;
                return new Restart(RestartType.ThrowIn, new Vec2(x, clampedY), opp);
            }

            // Goal line (outside the posts) → corner or goal kick.
            // Determine which goal line was crossed.
            bool crossedHomeGoalLine = ball.y <= -pitch.HalfLength; // Home defends -y
            TeamSide goalLineOwner = crossedHomeGoalLine ? TeamSide.Home : TeamSide.Away;

            float goalLineY = crossedHomeGoalLine ? -pitch.HalfLength : pitch.HalfLength;

            if (lastTouch == goalLineOwner)
            {
                // Defending team put it out → corner to the attackers.
                float cornerX = ball.x >= 0f ? pitch.HalfWidth : -pitch.HalfWidth;
                return new Restart(RestartType.Corner, new Vec2(cornerX, goalLineY), opp);
            }
            else
            {
                // Attackers put it out → goal kick to the defenders.
                float dir = crossedHomeGoalLine ? 1f : -1f;
                var spot = new Vec2(0f, goalLineY + dir * pitch.PenaltyAreaDepth * 0.5f);
                return new Restart(RestartType.GoalKick, spot, goalLineOwner);
            }
        }

        /// <summary>
        /// Awards a free kick (or penalty if inside the box) to <paramref name="attackingSide"/>
        /// at the foul location.
        /// </summary>
        public static Restart AwardFreeKick(Pitch pitch, Vec2 foulSpot, TeamSide attackingSide)
        {
            float defendingGoalLine = pitch.DefendingGoalLine(Other(attackingSide));
            // The foul is committed by the defenders, so the relevant box is the
            // one the attacking side is shooting at = defenders' goal line.
            float targetGoalLine = pitch.AttackingGoalLine(attackingSide);
            if (pitch.IsInPenaltyArea(foulSpot, targetGoalLine))
                return new Restart(RestartType.Penalty, pitch.AttackingPenaltySpot(attackingSide), attackingSide);

            return new Restart(RestartType.FreeKick, pitch.Clamp(foulSpot), attackingSide);
        }

        /// <summary>
        /// Offside judgement at the moment of a pass: the receiver is offside if,
        /// in the attacking half, they are beyond both the ball and the
        /// second-last defender (here approximated by the last outfield defender
        /// plus keeper) towards the opponent goal line.
        /// </summary>
        public static bool IsOffside(Pitch pitch, TeamSide attackingSide,
            Vec2 receiver, Vec2 ball, IReadOnlyList<MatchPlayer> defenders)
        {
            float oppGoal = pitch.AttackingGoalLine(attackingSide);
            float sign = oppGoal > 0f ? 1f : -1f; // +1 if attacking +y

            // Only applies in the attacking half.
            if (receiver.y * sign <= 0f) return false;

            // Receiver must be ahead of the ball to be considered.
            if (receiver.y * sign <= ball.y * sign) return false;

            // Find the two deepest defenders (largest y towards their own goal).
            float deepest = float.NegativeInfinity, secondDeepest = float.NegativeInfinity;
            for (int i = 0; i < defenders.Count; i++)
            {
                float d = defenders[i].Position.y * sign;
                if (d > deepest)
                {
                    secondDeepest = deepest;
                    deepest = d;
                }
                else if (d > secondDeepest)
                {
                    secondDeepest = d;
                }
            }

            if (float.IsNegativeInfinity(secondDeepest)) return false;

            // Offside if receiver is beyond the second-last defender line.
            return receiver.y * sign > secondDeepest + 0.1f;
        }

        private static float Clamp(float v, float min, float max) =>
            v < min ? min : (v > max ? max : v);
    }
}

File: js/match/referee.js
// referee.js — laws of the game: clock/halves/stoppage, fouls + advantage,
// the warning → yellow → red escalation (and second-yellow = red), offside
// judged at the moment of the pass, and which restart a stoppage produces.
import { V2, clamp } from '../util/mathx.js';
import { PITCH, inBounds, attackZ } from './pitch.js';

export const RestartType = {
  None: 'None', KickOff: 'KickOff', ThrowIn: 'ThrowIn', GoalKick: 'GoalKick',
  Corner: 'Corner', FreeKick: 'FreeKick', Penalty: 'Penalty', GoalCelebration: 'Goal',
};

export class Referee {
  constructor(halfLengthSeconds) {
    this.halfLength = halfLengthSeconds;
    this.clock = 0;          // seconds elapsed in current half
    this.half = 1;
    this.stoppage = 0;
    this.warnings = {};      // playerId -> count of verbal warnings
    this.log = [];
  }

  emit(msg) {
    this.log.push(`${this.timeLabel()}  ${msg}`);
    if (this.log.length > 60) this.log.shift();
  }

  timeLabel() {
    const base = this.half === 2 ? this.halfLength : 0;
    const total = base + this.clock;
    const m = Math.floor(total / 60);
    const s = Math.floor(total % 60);
    return `${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`;
  }

  tick(dt) {
    this.clock += dt;
  }

  get halfOver() { return this.clock >= this.halfLength + this.stoppage; }
  get matchOver() { return this.half >= 2 && this.halfOver; }

  // Foul severity 0..1 -> card decision. Ordinary tackles that mistime the ball
  // are just free kicks (no card). Only genuinely rough/reckless challenges
  // (high severity — typically a bad slide) are cautioned or sent off.
  judgeFoul(fouler, severity) {
    this.warnings[fouler.id] = this.warnings[fouler.id] || 0;
    let card = 'none';
    if (severity > 0.85) card = 'red';
    else if (severity > 0.6) card = 'yellow';
    else card = 'none'; // clean-but-mistimed challenge: free kick only

    if (card === 'yellow') {
      fouler.cardState = (fouler.cardState || 0) + 1;
      if (fouler.cardState >= 2) {
        card = 'second-yellow-red';
        fouler.sentOff = true;
        this.emit(`🟨🟥 Second yellow — ${fouler.profile.name} is SENT OFF!`);
      } else {
        this.emit(`🟨 Yellow card — ${fouler.profile.name}.`);
      }
    } else if (card === 'red') {
      fouler.cardState = 3;
      fouler.sentOff = true;
      this.emit(`🟥 Straight red — ${fouler.profile.name} is SENT OFF!`);
    }
    return card;
  }

  // Offside: attacker beyond second-last defender AND the ball at moment of pass.
  static isOffside(receiver, world) {
    const side = receiver.side;
    const dir = attackZ(side);
    const defenders = world.players.filter((q) => q.side !== side && !q.sentOff);
    // Find second-last defender line (includes GK).
    const zs = defenders.map((d) => d.pos.z * dir).sort((a, b) => b - a);
    if (zs.length < 2) return false;
    const secondLast = zs[1];
    const recv = receiver.pos.z * dir;
    const ballZ = world.ball.pos.z * dir;
    // Onside if level/behind second-last defender or not beyond the ball.
    if (recv <= secondLast + 0.4) return false;
    if (recv <= ballZ + 0.4) return false;
    // Only in the attacking half.
    if (receiver.pos.z * dir <= 0) return false;
    return true;
  }
}

// Decide the restart when the ball leaves the field. lastTouchSide is the side
// of the player who touched it last.
export function restartForOutOfPlay(ballPos, lastTouchSide) {
  const hw = PITCH.halfWidth, hl = PITCH.halfLength;
  if (Math.abs(ballPos.x) > hw) {
    return { type: RestartType.ThrowIn, side: lastTouchSide === 'home' ? 'away' : 'home',
             at: new V2(Math.sign(ballPos.x) * hw, clamp(ballPos.z, -hl + 2, hl - 2)) };
  }
  if (Math.abs(ballPos.z) > hl) {
    const endZ = Math.sign(ballPos.z) * hl;
    // Did the ball go off the end the 'home' or 'away' team defends?
    const defendingSide = endZ < 0 ? 'home' : 'away';
    if (lastTouchSide === defendingSide) {
      // Attacker last touched it going over their opponent's line => corner.
      return { type: RestartType.Corner, side: defendingSide === 'home' ? 'away' : 'home',
               at: new V2(Math.sign(ballPos.x || 1) * (hw - 0.5), endZ) };
    }
    // Attacker put it out => goal kick to the defending side.
    return { type: RestartType.GoalKick, side: defendingSide,
             at: new V2(Math.sign(ballPos.x || 1) * 8, endZ + (defendingSide === 'home' ? 5.5 : -5.5)) };
  }
  return { type: RestartType.None };
}
